sbtools data discovery

lindsaycarr opened the pull request:
Some of the `query_sb` function examples need some work (can't get spatial ones to work, and doi ones are causing issues). This fixes #113

lindsaycarr commented:
@aappling-usgs review/suggest edits once you return from travel!

aappling-usgs commented on content/usgs-packages/sbtools_Discovery.Rmd:
```diff
@@ -0,0 +1,224 @@
+
+# query by bounding box
+bbox_sp_obj <- sp::SpatialPoints(appalachia)
+# query_sb_spatial(bbox=bbox_sp_obj)
```
I ran a query by interactively selecting a bounding box under "Browse by Location" at https://www.sciencebase.gov/catalog/. The resulting URL was
```
https://www.sciencebase.gov/catalog/items?filter=spatialQuery%3D%7Btype%3A%22envelope%22%2Ccoordinates%3A%5B%5B-91.06475439079985%2C39.171001192353756%5D%2C+%5B-81.22100439080239%2C33.21851920313964%5D%5D%7D
```
which shares the `spatialQuery` filter with `query_sb_spatial`, but otherwise looks different from the WKT format that's ultimately used by all calls to `query_sb_spatial` (https://github.com/USGS-R/sbtools/blob/master/R/query_sb_spatial.R#L59). do you suppose SB might have changed formats? i don't see reference to WKT in their spatial docs.

lindsaycarr replied:
How would you figure out the projection we would need? Can we just pick the `proj4string`? 

And yes, I get an empty list for all of these queries.

Hmmm it does look quite different. I guess it might have changed? Should we make an sbtools issue about spatial queries?

aappling-usgs replied:
i know pitifully little about projections. maybe ask laura or a jordan?

yeah, if you're getting empty lists too, i guess another sbtools issue is in order.

aappling-usgs commented on content/usgs-packages/sbtools_Discovery.Rmd:
```diff
@@ -0,0 +1,224 @@
+# query by WKT polygon
+wkt_coord_str <- paste(conus$long, conus$lat, sep=" ", collapse = ",")
+wkt_str <- sprintf("POLYGON((%s))", wkt_coord_str)
+# query_sb_spatial(bb_wkt = wkt_str)
```
this runs for me, just returns an unrewarding `list()`. is that what you're getting, too?

aappling-usgs commented on content/usgs-packages/sbtools_Discovery.Rmd:
````diff
@@ -0,0 +1,224 @@
+```{r}
+# find data worked on in the last week
+today <- Sys.time()
+oneweekago <- today - (7*24*3600) # days * hrs/day * secs/hr
````
any interest in replacing this with
```r
oneweekago <- today - as.difftime(7, units='days')
```
?

aappling-usgs commented on content/usgs-packages/sbtools_Discovery.Rmd:
```diff
@@ -0,0 +1,224 @@
+# find data worked on in the last week
+today <- Sys.time()
+oneweekago <- today - (7*24*3600) # days * hrs/day * secs/hr
+recent_data <- query_sb_date(start = today, end = oneweekago)
```
i started by wondering whether it was OK for start to come after end, then happened on this curious behavior:
```r
> recent_data <- query_sb_date(start = today, end = oneweekago, limit=10000)
> length(recent_data)
[1] 114
> recent_data <- query_sb_date(end = today, start = oneweekago, limit=10000)
> length(recent_data)
[1] 122
```
...i don't know what to make of this. datetimes do get converted to dates (https://github.com/USGS-R/sbtools/blob/master/R/query_sb_date.R#L36), maybe one of start/end is inclusive while the other is exclusive?

i guess i'd recommend demonstrating with dates rather than datetimes, given that query_sb_date simplifies to dates no matter what. the inclusive/exclusive thing is curious but probably not worth teaching here.

lindsaycarr replied:
Tried doing this with Dates instead of datetimes, but still seeing this curious behavior...

```
length(query_sb_date(start = today, end = oneweekago, limit=10000))
[1] 10000
> length(query_sb_date(end = today, start = oneweekago, limit=10000))
[1] 407
> length(query_sb_date(start = as.Date(today), end = as.Date(oneweekago), limit=10000))
[1] 10000
> length(query_sb_date(end = as.Date(today), start = as.Date(oneweekago), limit=10000))
[1] 407
```

aappling-usgs replied:
wow, that's a much bigger difference than I saw! i guess it's possible that >=9593 items were created exactly on `as.Date(oneweekago)`...or that when end < start, strange things happen. couldn't say which without more testing, which i'm not sure is worth our time.

that said, i didn't actually expect the Date approach to change the behavior - just to help clarify that Dates are the best precision you can hope for

lindsaycarr replied:
Yeah, that seems like something we don't need to defend against here.

Ah I see your point in doing that. I'll update it.

aappling-usgs commented on content/usgs-packages/sbtools_Discovery.Rmd:
```diff
@@ -0,0 +1,224 @@
+
+### Using `query_sb_datatype`
+
+`query_sb_datatype` is used to search ScienceBase by the type of data an item is listed as. Run `sb_datatypes()` to get a list of 50 available data types.
```
ooh, good tip! and looks like this is also the list of possible `browseType`s? i see that `query_sb_datatype` creates a `browseType` filter, and element 46 ("Static Map Image") is what you used for `browseType` above in line 81...if you do move `query_sb()` down to below all these more specific `query_sb_...()`s, you could just refer back to this function when you do the `query_sb(browseType...)` example

aappling-usgs commented on content/usgs-packages/sbtools_Discovery.Rmd:
````diff
@@ -0,0 +1,224 @@
+})
+```
+
+## No results
````
maybe this should go above Best of Both Methods, given that it refers to the first method only?

aappling-usgs commented on content/usgs-packages/sbtools_Discovery.Rmd:
````diff
@@ -0,0 +1,224 @@
+
+```{r}
+# search for items related to a Water Quality Portal paper DOI
+query_results <- query_sb_doi(doi = '10.1002/2016WR019993')
````
lol. making good use of a dead end from your work on the `query_sb_doi` section, eh? there's no chance that this paper will ever make its way onto SB, is there?

lindsaycarr replied:
Not that I know of!

lindsaycarr commented on content/usgs-packages/sbtools_Discovery.Rmd:
```diff
@@ -0,0 +1,224 @@
+
+# search by keyword
+precip_query <- list(q = 'precipitation')
+precip_data <- query_sb(query_list = precip_query)
```
I don't think I really had a reason...I think I just did it. Thanks for pointing it out, I agree - no need to clutter the environment for something that is used once.

lindsaycarr commented on content/usgs-packages/sbtools_Discovery.Rmd:
````diff
@@ -0,0 +1,224 @@
+
+```{r}
+
+##### THESE FIRST TWO SEARCHES STILL NEED WORK #####
````
I think I was trying out a few other query args, like sort or something and wasn't getting what I expected. I just tried adding `sort` to the query list and it came out just fine. Consider "still need work" to no longer be true.

lindsaycarr commented on content/usgs-packages/sbtools_Discovery.Rmd:
```diff
@@ -0,0 +1,224 @@
+head(precip_data, 2)
+
+# search by keyword + category
+precip_maps_query <- list(q = 'precipitation', browseType = "Static Map Image", sort='title')
```
Yeah, I was not sure what that meant but `sb_datatypes()` gave me the list for `browseType`. For `browseCategory`, I went to SB, and it didn't look like there were any more categories (https://www.sciencebase.gov/catalog/).

lindsaycarr commented on content/usgs-packages/sbtools_Discovery.Rmd:
```diff
@@ -0,0 +1,224 @@
+##### -------------------------------------- #####
+
+# search by 2 keywords
+hazard2_query <- list(q = 'flood', q = 'earthquake')
```
Since the behavior of `q` when given multiple keywords seems unexpected, I'm just going to replace those examples with Lucene query examples instead.

aappling-usgs replied:
sounds good to me!

lindsaycarr commented on content/usgs-packages/sbtools_Discovery.Rmd:
````diff
@@ -0,0 +1,224 @@
+length(hazard3_data)
+head(hazard3_data, 2)
+```
+
````
Struggling with what that actually looks like via sbtools:

```r
length(query_sb(query_list = list(q = '', lq = 'flood AND earthquake', browserType = "!Application"), limit=200))
```

aappling-usgs replied:
hmm, me neither. I tried both `'browseCategory!'='Image'` and `'browseCategory'='!Image'`, both of which put the `!` in the right place, but then `!` gets converted to `%21` in `httr::handle_url` within `query_sb`, so it's all for naught. I made an `sbtools` issue to document the wished-for feature: https://github.com/USGS-R/sbtools/issues/238. Maybe for now we should just not introduce this topic in these lessons.

lindsaycarr replied:
Thought it might need to be escaped, but that didn't seem to make any difference. Sounds good, I will leave it alone for now!

lindsaycarr commented on content/usgs-packages/sbtools_Discovery.Rmd:
```diff
@@ -0,0 +1,224 @@
+head(park_results, 2)
+
+# search using a site location
+loc_results <- query_sb_text("Embudo")
```
Hmmm that's very true. All of these get handled in Science Base the same way I think, right? Since it's just searching text. Could do `river_results <- query_sb_text("Rio Grande")` to show another slightly different example.

aappling-usgs replied:
sure, either Embudo or Rio Grande is fine if you're just looking for several examples for reinforcement. Rio Grande does seem a touch more different.

lindsaycarr commented on content/usgs-packages/sbtools_Discovery.Rmd:
```diff
@@ -0,0 +1,224 @@
+
+### Using `query_sb_spatial`
+
+`query_sb_spatial` accepts 3 different methods for specifying a spatial area in which to look for data. You can specify a bounding box `bbox` as an `sp` spatial data object *[[[[NEED MORE TEXT HERE]]]]*.  Alternatively, you can supply a vector of latitudes and a vector of longitudes using `lat` and `long` arguments. The function will automatically use the minimum and maximum from those vectors to ...
```
the `MORE TEXT` tag was because I thought I needed to explain what an sp object is. Or at least link out to something that describes it

aappling-usgs replied:
i like the link-out idea

lindsaycarr commented on content/usgs-packages/sbtools_Discovery.Rmd:
```diff
@@ -0,0 +1,224 @@
+lapply(sbmaps, function(sbitem) {
+  url <- sbitem[['link']][['url']]
+  browseURL(url)
+  return(url)
```
Was thinking that it would be nice to just print out the URLs that were opened. I don't know why, but I generally don't like functions that don't return anything.

aappling-usgs replied:
ok, fine with me.

lindsaycarr commented:
Updated everything based on @aappling-usgs feedback. There is a known issue with spatial data queries returning empty lists (https://github.com/USGS-R/sbtools/issues/237), so that will need to be revisited and updated before this curriculum "goes live". I've made an issue here for that - #270.

aappling-usgs commented on content/usgs-packages/sbtools_Discovery.Rmd:
```diff
@@ -212,13 +197,67 @@ lapply(sbmaps, function(sbitem) {
+length(atlantic_ocean)
+head(atlantic_ocean)
+
+# date query during Marco Polo's life
```
:smiley_cat:
